Added Jupyter console widget and Example

jonmatthis opened the pull request:
I added a JupyterConsole widget and made an `Example` showing basic use

this was based on the implementation from `yaqc-qtpy` which I learned about in a conversation on the `#pyqtgraph` channel in the Python discord: https://discord.com/channels/267624335836053506/898139460821192724/991360299158470676


I did my best to follow the `CONTRIBUTING.md` guidelines, but I'm relatively new at this so I might have missed something

j9ac9k commented on pyqtgraph/widgets/JupyterConsoleWidget.py:
```diff
@@ -0,0 +1,45 @@
+        self.addTab(self.rich_jupyter_widget, "console")
+
+    def execute_command(self,command_string:str=None)->bool:
+            return self.rich_jupyter_widget.execute(command_string)
```
you may have mixed tabs and spaces here?

jonmatthis replied:
That's part of the code I got from `yaqc` so i guess I can blame them 😅 

I formatted the files with `black`, so it should be good now

j9ac9k commented:
Hi @jonmatthis 

Thanks so much for this PR.  I always thought pyqtgraph's console example probably wasn't that great and we would better serve our users by showing how to better integrate jupyter consoles or the qtconsole.

A couple of small issues I see, seems like the formatting isn't quite right, I think you may have a tab instead of spaces in one place and such; you may want to run`black` on the new-file.

Also in `pyqtgraph/examples/utils.py` you reference a file that doesn't exist, but in this PR you add the file to `pyqtgraph/widgets/...`. I'm not sure if you forgot to include the example file, or if the file in the widgets directory should be in the examples.

ntjess commented:
Nifty! I imagine updating the kernel's namespace would be a fairly common operation; would you consider adding a method similar to `pushVariables` from [this SO answer](https://stackoverflow.com/questions/11513132/embedding-ipython-qt-console-in-a-pyqt-application/20610786#20610786)? I would also support adding a `namespace` argument to the constructor for this reason, like the old [`ConsoleWidget`](https://pyqtgraph.readthedocs.io/en/latest/widgets/consolewidget.html?highlight=namespace)

jonmatthis commented:
Whoops, I forgot to add the `example`  file to git. I think I just did that? 

I'll go through your other comments now, thanks!

jonmatthis commented:
> Nifty! I imagine updating the kernel's namespace would be a fairly common operation; would you consider adding a method similar to `pushVariables` from [this SO answer](https://stackoverflow.com/questions/11513132/embedding-ipython-qt-console-in-a-pyqt-application/20610786#20610786)?

I agree that's very useful! I added that method and showed and sample usage in the `example`

> I would also support adding a `namespace` argument to the constructor for this reason, like the old [`ConsoleWidget`](https://pyqtgraph.readthedocs.io/en/latest/widgets/consolewidget.html?highlight=namespace)

I did this too, I'm not 100% clear on the concept of a 'namespace' so it's possible I missed something important :)

jonmatthis commented on pyqtgraph/examples/jupyterConsoleWidgetExample.py:
```diff
@@ -0,0 +1,39 @@
+ This widget was adapted from the implementation from - https://gitlab.com/yaq/yaqc-qtpy/-/blob/main/yaqc_qtpy/_main_widget.py
+
+"""
+import time
```
fixed

j9ac9k commented:
Hi @jonmatthis 

Thanks for updating the PR.  I'm reluctant to add code here to pyqtgraph/widgets, I think this is small enough, the entirety can live as an example.

In the discord chat, @pijyoi was kind enough to make a suggestion for what an example could look like:

```python
import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtWidgets
from qtconsole import inprocess

class ConsoleWidget(inprocess.QtInProcessRichJupyterWidget):
    def __init__(self):
        super().__init__()
        
        self.kernel_manager = inprocess.QtInProcessKernelManager()
        self.kernel_manager.start_kernel()
        self.kernel_client = self.kernel_manager.client()
        self.kernel_client.start_channels()

    def shutdown_kernel(self):
        self.kernel_client.stop_channels()
        self.kernel_manager.shutdown_kernel()

class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()
        cwidget = QtWidgets.QWidget()
        vbox = QtWidgets.QVBoxLayout()
        cwidget.setLayout(vbox)
        self.plot = pg.PlotWidget()
        self.console = ConsoleWidget()
        vbox.addWidget(self.plot)
        vbox.addWidget(self.console)
        self.setCentralWidget(cwidget)

        app = QtWidgets.QApplication.instance()
        app.aboutToQuit.connect(self.console.shutdown_kernel)

        kernel = self.console.kernel_manager.kernel
        kernel.shell.push(dict(np=np, pw=self.plot))
        self.console.execute("whos")

if __name__ == '__main__':
    pg.mkQApp()
    main = MainWindow()
    main.show()
    pg.exec()
```

There is some president here for demonstrating functionality in the examples that isn't really a part of the library directly, specifically [the GLPainter example](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/GLPainterItem.py) comes to mind.

jonmatthis commented:
I agree with this! It demonstrates the method to get a jupyter console widget, while keeping responsibility for maintaining that core functionality of that widget in the hands of the `qtconsole` folks, which is more appropriate

jonmatthis commented:
A few suggested tweaks to that example to make it a bit more transparent to low-XP folks (like me😅). 

It's mostly just changing the variable names to make it (even) more obvious which variable is what, and an example of how to execute a command after creating the Widget. 


Other potential cool additions would be -
- show sample inputs to plot something simple in that `Plot` window 
- Adding a very simple `paramter_tree` widget that shows the variables returned by `whos`


```python 
import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtWidgets
from qtconsole import inprocess


class JupyterConsoleWidget(inprocess.QtInProcessRichJupyterWidget):
    def __init__(self):
        super().__init__()

        self.kernel_manager = inprocess.QtInProcessKernelManager()
        self.kernel_manager.start_kernel()
        self.kernel_client = self.kernel_manager.client()
        self.kernel_client.start_channels()

    def shutdown_kernel(self):
        self.kernel_client.stop_channels()
        self.kernel_manager.shutdown_kernel()


class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()
        central_widget = QtWidgets.QWidget()
        vbox_widget = QtWidgets.QVBoxLayout()
        central_widget.setLayout(vbox_widget)
        self.plot_widget = pg.PlotWidget()
        self.console = JupyterConsoleWidget()
        vbox_widget.addWidget(self.plot_widget)
        vbox_widget.addWidget(self.console)
        self.setCentralWidget(central_widget)

        app = QtWidgets.QApplication.instance()
        app.aboutToQuit.connect(self.console.shutdown_kernel)

        kernel = self.console.kernel_manager.kernel
        kernel.shell.push(dict(np=np, pw=self.plot_widget))
        self.console.execute("whos")


if __name__ == '__main__':
    pg.mkQApp()
    main = MainWindow()
    main.show()
    main.console.execute("print(\"hello world :D \")")
    pg.exec()
```

j9ac9k commented:
<img width="810" alt="image" src="https://user-images.githubusercontent.com/646398/177062072-026edb07-bf95-48b4-8b14-92a33f97d319.png">

Nice, in the PR there was an example of how to incorporate dark mode; can that be easily done?

pijyoi commented:
> in the PR there was an example of how to incorporate dark mode; can that be easily done?

Does the following work?

I also renamed "vbox_widget" to "vbox_layout".

```python
import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtWidgets
from qtconsole import inprocess


class JupyterConsoleWidget(inprocess.QtInProcessRichJupyterWidget):
    def __init__(self):
        super().__init__()

        self.kernel_manager = inprocess.QtInProcessKernelManager()
        self.kernel_manager.start_kernel()
        self.kernel_client = self.kernel_manager.client()
        self.kernel_client.start_channels()

    def shutdown_kernel(self):
        self.kernel_client.stop_channels()
        self.kernel_manager.shutdown_kernel()


class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()
        central_widget = QtWidgets.QWidget()
        vbox_layout = QtWidgets.QVBoxLayout()
        central_widget.setLayout(vbox_layout)
        self.plot_widget = pg.PlotWidget()
        self.console = JupyterConsoleWidget()
        vbox_layout.addWidget(self.plot_widget)
        vbox_layout.addWidget(self.console)
        self.setCentralWidget(central_widget)

        app = QtWidgets.QApplication.instance()
        app.aboutToQuit.connect(self.console.shutdown_kernel)
        app.paletteChanged.connect(self.onPaletteChanged)
        self.onPaletteChanged(None)

        kernel = self.console.kernel_manager.kernel
        kernel.shell.push(dict(np=np, pw=self.plot_widget))
        self.console.execute("whos")

    def onPaletteChanged(self, palette):
        darkMode = QtWidgets.QApplication.instance().property("darkMode")
        style = ("lightbg", "linux")[darkMode]
        self.console.set_default_style(style)

if __name__ == '__main__':
    pg.mkQApp()
    main = MainWindow()
    main.show()
    main.console.execute("print(\"hello world :D \")")
    pg.exec()
```

jonmatthis commented:
sorry for the delay! I'll test this on my end later today, but I suspect that I'll be totally happy with this option :D

jonmatthis commented:
Ok, I *think* I removed my previously committed changes (but please double check I didn't miss anything) 

To summarize changes: 

Per the conversation above, I removed the standalong "JupyterConsoleWidget" in favor of a more fleshed out `example` that implements such a console based on the `qtconsole` project's examples laid out here: https://github.com/jupyter/qtconsole/tree/b4e08f763ef1334d3560d8dac1d7f9095859545a/examples

The new example creates a `RichJupyterConsole` Widget alongside a `PlotWidget`, and then shows an example of a way to plot simple sine/cosine waves by entering commands from the `__main__` function of the example. 

The widgets are implements as Docks so that Users can re-size the windows and close the PlotWidget if they want to. 

The example was added to the list of examples in `examples/utils.py`

I tagged it as a 'recommended' example, but I'm totally happy to remove that if that doesn't feel appropriate. 

Please let me know if there is anything else I can do to help move this forward :D

j9ac9k commented on .gitignore:
```diff
@@ -110,3 +110,4 @@ asv.conf.json
 
 # jupyter notebooks
 .ipynb_checkpoints
+.venv
```
Please insert new-line 👍🏻

jonmatthis replied:
ok, done!

j9ac9k replied:
haha no, a newline _after_ the `.venv` :)

jonmatthis replied:
ooooh, that makes more sense 😅

j9ac9k commented:
Hi @jonmatthis 

Thanks for updating the PR, sorry it's taken me a bit to get to this.

I added an inline comment about inserting a newline in `.gitignore`; the other thing is I think you need to remove `pyqtgraph/examples/jupyterConsoleWidgetExample.py` as that used the JupyterConsoleWidget class in pyqtgraph.widgets which is no longer present.

Also the dark mode is 🔥

jonmatthis commented:
Ok, done!

I knew I was forgetting something 😅

j9ac9k commented on pyqtgraph/examples/jupyter_console_example.py:
```diff
@@ -0,0 +1,97 @@
+
+try:
+    from qtconsole import inprocess
+except:
```
specify the exception to catch as being `ImportError` here, not just a generic exception.

jonmatthis commented:
Ok, did the things 👍

pijyoi commented on pyqtgraph/examples/jupyter_console_example.py:
```diff
@@ -23,7 +23,7 @@
 try:
     from qtconsole import inprocess
-except:
+except ImportError or NameError:
```
I don't think the `or` operator does what you want it to.
```python
class A(Exception): pass
class B(Exception): pass

try:
    raise B()
except A or B:
    print('caught')
except:
    print('oops')
```

j9ac9k replied:
Ahh, correct, what is needed here is `except (ImportError, NameError):

https://docs.python.org/3/tutorial/errors.html#handling-exceptions

jonmatthis replied:
done!

pijyoi commented:
Do we want to add this example into `conditionalExamples` in https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/test_examples.py#L68 in order to avoid a CI dependency on `qtconsole`?

j9ac9k commented:
> Do we want to add this example into `conditionalExamples` in https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/test_examples.py#L68 in order to avoid a CI dependency on `qtconsole`?

Completely forgot about `conditionalExamples`, thanks for the reminder @pijyoi this would certainly qualify.

j9ac9k commented:
blerg, need to indicate to skip this example when qtconsole is not installed; sorry should have caught this earlier sorry about that!

Anyway, you need to add an entry here:

https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/test_examples.py#L68-L77

You'll want to add something like

```python
    "jupyter_console_example.py": exceptionCondition(
        importlib.util.find_spec("qtconsole") is not None,
        reason="No need to test with qtconsole not being installed"
    ),
```

j9ac9k commented:
Hi @jonmatthis I hope you don't mind, I added that last bit to make the examples pass; I should have given you that feedback weeks ago, sorry about that.  Anyway, this looks good to merge.  Thanks for the PR!

jonmatthis commented:
Hey cool!! This is my first PR into an open source project :D

Thank you for all your great work maintaining this incredible package, and for making this PR experience easy, fun, and educational <3

j9ac9k commented:
Congratulations on your first PR!  Appreciate the sentiment, we certainly could have done better by giving you reviews/suggestions in a more timely manner, sorry about that!
